### ISSUE TITLE
Timestamps in public beta

### ISSUE THREAD

**Issue by frankmcsherry** [MaterializeInc/materialize#1309]
Event vs System time

tl;dr

**Right now, we are not respecting event time (we fast-forward out-of-order Kafka timestamps), and we are not reflecting system time (we hold back data to line up with oldest timestamps)**. 

I propose that in the scope of the public beta, we actively violate event time, in favor of system time, but do so publicly and clearly. Going forward, post-beta, I think we should use multitemporal timestamps to provide an interactive experience without losing any of the precision of event time.

Main action item: make flush-tables.sh happen more frequently, make peeks happen at non-decreasing times. 

---

Right now we have a single integer for timestamps. I think we'll need to stick with that for the beta and perhaps early engagements. There are more clever and complicated things we can do with differential (keyword "multitemporal"), but there are a lot of things to shake about wrt actual humans understanding things.

The two common frames of reference for times are "event time" and "system time". We probably want to choose one, or conflate the two somehow. Ideally, clearly and intentionally, which I think we aren't exactly doing right now.

Event time refers to times described in the source data, for example the Kafka timestamps. Lots of folks think that you should reflect these times, because the data producer probably wrote them down for a good reason.

System time refers to the times experienced by materialized, or an analyst using materialized. Lots of folks think that you should respect these times, because the analyst is the one driving the computation around and they want to know about their "now", rather than having information held back from them.

These times are in conflict, sometimes inherently (data can be late, historical corrections must be made, etc) and sometimes because systems are bad (e.g. that no one bothers to say that `nations` hasn't changed, second by second). We want to reconcile these issues, at least in the public demo timeframe.

**Right now, we are not respecting event time (we fast-forward out-of-order Kafka timestamps), and we are not reflecting system time (we hold back data to line up with oldest timestamps)**. 

Doing neither isn't obviously *wrong*; it's hard to respect event time because we can easily stall if we do, and it's awkward to reflect system time because we don't want to hand out incorrect/inconsistent answers (i.e. we do want to hold back data we know about, until it matches the rest of the data coming in). If we tracked either event time or system time religiously, we would probably annoy the user.

---

I propose that in the scope of the public beta, we actively violate event time, in favor of system time, but do so publicly and clearly. Going forward, post-beta, I think we should use multitemporal timestamps to provide an interactive experience without losing any of the precision of event time.

Specifically, imagine a world where materialized clusters sources (into what we internally call "databases") and both i. tracks the most advanced time within a cluster, and ii. fast-forwards all sources to that most advanced time. In the case of Kafka, this might mean tracking the leading frontier of committed times from each Kafka source, and introducing watermark empty before/after records in all topics (the way flush-sources.sh does now). Peeks are served with the current trailing edge across the cluster of sources. New sources can be introduced whenever, but their use probably stalls the peek until we get to the first watermark element.

This hypothetical behavior ensures that each database moves forward through time monotonically, with the leading edge ensuring that each other topic is forced to catch up, and the trailing edge ensuring that queries i. return promptly, ii. are correct for some specific time.

The clustering requirement might be awkward (we think several demo use cases involve non-RDBMS changelog data), and we might want to adopt a system-wide clock (either real time as observed at the coordinator, or leading edge of datasets). This is underspecified, and tbh there is a bunch of work to do to spec out what semantics and durability one wants here (perhaps: if you want durability put your files in Kafka!).

**Comment by frankmcsherry** [comment]
I believe this issue is no longer relevant to the current timestamp issues.
